disasm-test/Main.hs

```diff
@@ -318,119 +361,200 @@ cube = TS.mkCUBE
-           askOption $ \keep ->
-           askOption $ \details ->
-           testCase pfx
-           $ flip evalStateT (TestState { keepTemp = keep
-                                        , rndTrip = roundtrip
-                                        , showDetails = details
-                                        , llvmAs = llvmAs'
-                                        , llvmDis = llvmDis'
-                                        })
-           $ with2Files (processLL pfx file) $ \(parsed1, ast) ->
+       let tmod = if skipTest
+                  then ignoreTestBecause "not valid for this LLVM version"
+                  else case lookup (TS.rootMatchName sweet) knownBugs of
+                         Just why -> expectFailBecause why
+                         Nothing -> id
+       let pfx = TS.rootBaseName sweet
+       return $ (:[]) $ tmod
+         $ testCaseM pfx
+         $ with2Files (processLL pfx $ TS.rootFile sweet)
+         $ \(parsed1, ast) ->
              case ast of
                Nothing   -> return ()
                Just ast1 ->
                  -- Re-assemble and re-disassemble
                  with2Files (processLL pfx parsed1) $ \(_, Just ast2) -> do
-                 diff ast1 ast2 -- Ensure that the ASTs match
-
-                 -- Ensure that the disassembled files match.
-                 -- This is usually too strict (and doesn't really provide more info).
-                 -- We normalize the AST (see below) to ensure that the ASTs match modulo
-                 -- metadata numbering, but the equivalent isn't possible for the
-                 -- assembly: we need llvm-as to be able to re-assemble it.
-                 -- diff parsed1 parsed2
-
-  where file  = TS.rootFile sweet
-        pfx   = TS.rootBaseName sweet
-        assertF ls = liftIO $ assertFailure $ unlines ls
-        diff file1 file2 = do
-          (code, stdout, stderr) <- liftIO $
-            Proc.readCreateProcessWithExitCode (Proc.proc "diff" ["-u", file1, file2]) ""
-          case code of
-            ExitFailure _ -> assertF ["diff failed", stdout, stderr]
-            ExitSuccess   ->
-              if stdout /= "" || stderr /= ""
-              then assertF ["non-empty diff", stdout, stderr]
-              else do Details dets <- gets showDetails
-                      when dets $ liftIO
-                        $ mapM_ putStrLn ["success: empty diff: ", file1, file2]
-
+                 diffCmp ast1 ast2 -- Ensure that the ASTs match
+
+                 -- Ensure that the disassembled files match.  This is usually
+                 -- too strict (and doesn't really provide more info).  We
+                 -- normalize the AST (see below) to ensure that the ASTs match
+                 -- modulo metadata numbering, but the equivalent isn't possible
+                 -- for the assembly: we need llvm-as to be able to re-assemble
+                 -- it.
+                 --
+                 -- diffCmp parsed1 parsed2
```

Comment:
Is this a stray comment? There is another use of `diffCmp` directly above this, and I'm not sure if the comment here is meant to apply to that line or not.

Reply:
It is intentional: it's talking about comparing  the .ll files (represented by `parsed1` and `parsed2`), as opposed to the serialized AST (via `ppLLVM`) above that.